datafusion/datasource/src/memory.rs

```diff
@@ -902,4 +1108,319 @@ mod tests {
+        )?;
+        assert_partitioning(partitioned_datasrc, Some(5));
+
+        // src has partitions < target partitions, and can split to sufficient amount
+        // has 6 batches across 3 partitions. Will need to split all of it's partitions.
+        let target_partitions = 6;
+        let mem_src_config = memorysrcconfig_3_partitions_with_2_batches_each(
+            sort_information_on_config.clone(),
+        )?;
+        let partitioned_datasrc = mem_src_config.repartitioned(
+            target_partitions,
+            not_used,
+            output_ordering.clone(),
+        )?;
+        assert_partitioning(partitioned_datasrc, Some(6));
+
+        // src has partitions < target partitions, but not enough total batches to fulfill the split (desired target_partitions)
+        let target_partitions = 3 * 2 + 1;
+        let mem_src_config = memorysrcconfig_3_partitions_with_2_batches_each(
+            sort_information_on_config.clone(),
+        )?;
+        let partitioned_datasrc = mem_src_config.repartitioned(
+            target_partitions,
+            not_used,
+            output_ordering.clone(),
+        )?;
+        assert_partitioning(partitioned_datasrc, None);
+
+        // src has 1 partition with many batches of lopsided sizes
+        // make sure it handles the split properly
+        let target_partitions = 2;
+        let mem_src_config = memorysrcconfig_1_partition_with_different_sized_batches(
+            sort_information_on_config,
+        )?;
+        let partitioned_datasrc = mem_src_config.clone().repartitioned(
+            target_partitions,
+            not_used,
+            output_ordering,
+        )?;
+        assert_partitioning(partitioned_datasrc.clone(), Some(2));
+        // Starting = batch(100_000), batch(10_000), batch(100), batch(1).
+        // It should have split as p1=batch(100_000), p2=[batch(10_000), batch(100), batch(1)]
+        let repartitioned_raw_batches = mem_src_config
+            .repartition_evenly_by_size(target_partitions)?
+            .unwrap();
+        assert_eq!(repartitioned_raw_batches.len(), 2);
+        let [ref p1, ref p2] = repartitioned_raw_batches[..] else {
+            unreachable!()
+        };
+        // p1=batch(100_000)
+        assert_eq!(p1.len(), 1);
+        assert_eq!(p1[0].num_rows(), 100_000);
+        // p2=[batch(10_000), batch(100), batch(1)]
+        assert_eq!(p2.len(), 3);
+        assert_eq!(p2[0].num_rows(), 10_000);
+        assert_eq!(p2[1].num_rows(), 100);
+        assert_eq!(p2[2].num_rows(), 1);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_repartition_no_sort_information_no_output_ordering() -> Result<()> {
+        let no_sort = vec![];
+        let no_output_ordering = None;
+
+        // Test: common set of functionality
+        run_all_test_scenarios(no_output_ordering.clone(), no_sort.clone())?;
+
+        // Test: does not preserve separate partitions (with own internal ordering) on even split
```

Comment:
This defines what I believe the different behavior should be when not preserving sort ordering. I looked at how [FileScanConfig::repartition](https://github.com/apache/datafusion/blob/908ca0e684f3184075ef799128ee3f58f1e9db86/datafusion/datasource/src/file_scan_config.rs#L237) was implemented as a guide. The key branch point in it's implementation is here:

https://github.com/apache/datafusion/blob/908ca0e684f3184075ef799128ee3f58f1e9db86/datafusion/datasource/src/file_groups.rs#L194-L199